xdsl/irdl/attributes.py

```diff
@@ -81,12 +83,56 @@ def generic_constraint_coercion(args: tuple[Any]) -> AttrConstraint:
 
 _A = TypeVar("_A", bound=Attribute)
 
+
+# Will deprecate soon
 ParameterDef = Annotated[_A, IRDLAnnotations.ParamDefAnnot]
 
 
-def irdl_param_attr_get_param_type_hints(cls: type[_A]) -> list[tuple[str, Any]]:
+# Field definition classes for `@irdl_param_attr_definition`
+class _ParameterDef:
+    param: AttrConstraint | Attribute | type[Attribute] | TypeVar | ConstraintVar
+
+    def __init__(
+        self,
+        param: AttrConstraint | Attribute | type[Attribute] | TypeVar | ConstraintVar,
+    ):
+        self.param = param
+
+
+@overload
+def param_def(
+    constraint: TypeVar,
+    *,
+    default: None = None,
+    resolver: None = None,
+    init: Literal[False] = False,
+) -> Attribute: ...
```

Comment:
This doesn't feel ideal? We effectively lose the python typing for the field which was kind of the whole point of having the typevar

Reply:
yep and it's currently breaking my code when I migrate xDSL to it. I'll probably change this soon.